pymc3/distributions/simulator.py

```diff
@@ -14,112 +14,238 @@
+                #     sum_stat = "identity"
+            elif callable(distance):
+                distance = create_distance_op_from_fn(distance)
+            else:
+                raise ValueError(f"The distance metric {distance} is not implemented")
+
+        if not isinstance(sum_stat, Op):
+            if sum_stat == "identity":
+                sum_stat = identity
+            elif sum_stat == "sort":
+                sum_stat = at.sort
+            elif sum_stat == "mean":
+                sum_stat = at.mean
+            elif sum_stat == "median":
+                # Missing in Aesara, see aesara/issues/525
+                sum_stat = create_sum_stat_op_from_fn(np.median)
+            elif callable(sum_stat):
+                sum_stat = create_sum_stat_op_from_fn(sum_stat)
+            else:
+                raise ValueError(f"The summary statistic {sum_stat} is not implemented")
+
+        epsilon = at.as_tensor_variable(floatX(epsilon))
+
+        if params is None:
+            params = unnamed_params
         else:
-            raise ValueError(f"The summary statistics {sum_stat} is not implemented")
-
-        super().__init__(shape=np.prod(observed.shape), dtype=observed.dtype, *args, **kwargs)
-
-    def random(self, point=None, size=None):
-        """
-        Draw random values from Simulator.
-
-        Parameters
-        ----------
-        point: dict, optional
-            Dict of variable values on which random values are to be conditioned (uses default
-            point if not specified).
-        size: int, optional
-            Desired size of random sample (returns one sample if not specified).
-
-        Returns
-        -------
-        array
-        """
-        # size = to_tuple(size)
-        # params = draw_values([*self.params], point=point, size=size)
-        # if len(size) == 0:
-        #     return self.function(*params)
-        # else:
-        #     return np.array([self.function(*params) for _ in range(size[0])])
+            if unnamed_params:
+                raise ValueError("Cannot pass both unnamed parameters and `params`")
+
+        # Assume scalar ndims_params
+        if ndims_params is None:
+            ndims_params = [0] * len(params)
+
+        sim_op = type(
+            f"Simulator_{name}",
+            (SimulatorRV,),
+            dict(
+                name="Simulator",
+                ndim_supp=ndim_supp,
+                ndims_params=ndims_params,
+                dtype=dtype,
+                inplace=False,
+                fn=fn,
+                _distance=distance,
+                _sum_stat=sum_stat,
+                epsilon=epsilon,
+            ),
+        )()
+
+        # The logp function is registered to the more general SimulatorRV,
+        # in order to avoid issues with multiprocessing / pickling
+
+        # rv_type = type(sim_op)
+        # NoDistribution.register(rv_type)
+        NoDistribution.register(SimulatorRV)
+
+        # @_logp.register(rv_type)
```

Comment:
@lucianopaz, I cannot register it on the new `rv_type`, or I get a weird failure on multiprocessing, where the logp defaults to the standard `at.zeros_like(value)`. 

This causes the first Simulator test to fail, complaining that the input variables are not used in the compiled function (which is true because the default `logp` is based only on `value`)

 The strange thing, is that when I run it on Jupyter, it fails in the first time the cell is run but passes the second time.